### ISSUE TITLE
V53 Cryptographic requirements for WebRTC

### ISSUE THREAD

**Issue by randomstuff** [OWASP/ASVS#2412]
Spinoff of #2215.

Is there any need for cryptographic requirements for WebRTC channels?

Possible topics:

* forward secrecy;
* peer authentication;
* encryption method/ciphersuites.

I would not expect much verification would be needed in practice. Maybe when a backend server acts as a WebRTC peer?

**References**

[WebRTC requires Perfect Forward Secrecy (PFS) starting in Firefox 38](https://hacks.mozilla.org/2015/02/webrtc-requires-perfect-forward-secrecy-pfs-starting-in-firefox-38/) indicates that forward secrecy was not always enabled.

[WebRTC Security: What You Need to Know in 2021](https://www.liveswitch.io/blog/webrtc-security-guide-2021) seems to imply that it is (was) not always automatic.

[RFC8826](https://www.rfc-editor.org/rfc/rfc8826.html):

>  It is this consideration that makes an automatic, public key-based key exchange mechanism imperative for WebRTC (this is a good idea for any communications security system), and this mechanism SHOULD provide Forward Secrecy (FS). The signaling channel/calling service can be used to authenticate this mechanism.

[RFC8827](https://www.rfc-editor.org/rfc/rfc8827.html#name-communications-security):

>  Implementations MUST favor cipher suites which support Forward Secrecy (FS) over non-FS cipher suites and SHOULD favor Authenticated Encryption with Associated Data (AEAD) over non-AEAD cipher suites. .

**Comment by tghosth** [comment]
I think we should actually ping @sandrogauci about this :) Sandro do you have any thoughts on this issue?

**Comment by sandrogauci** [comment]
Hi @randomstuff, thanks for the question. Here's my take:

The cryptographic requirements for WebRTC's DTLS implementation align with standard TLS requirements. Therefore, existing cryptographic requirements can be referenced where needed. This includes Perfect Forward Secrecy (PFS) considerations as part of the cipher suite requirements.

We've already addressed this in ASVS through requirement [53.2.2](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x53-V53-WebRTC.md#v532-media), which uses intentionally broad terms like "strong" and "secure" to remain adaptable to evolving cryptographic standards. This requirement was included for two key reasons:

1. Server-side cryptographic vulnerabilities can exist in WebRTC media gateways, regardless of browser security capabilities
2. Testing DTLS in WebRTC requires different methodologies than traditional TLS/HTTPS testing (e.g., testssl.sh approaches aren't applicable)

We want to avoid duplicating the cryptographic requirements already covered in the [Cryptography chapter](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x14-V6-Cryptography.md).

Regarding peer authentication, WebRTC implements mutual authentication during DTLS key establishment. This process uses certificate fingerprints transmitted via SDP (signaling), often with self-signed certificates. Both parties verify these fingerprints against the certificates. While self-signed certificates are typically considered insecure in other contexts, they are acceptable in WebRTC's security model. We address this in requirement [53.2.8](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x53-V53-WebRTC.md#v532-media), which ensures proper fingerprint verification.

Does this answer your question?

**Comment by randomstuff** [comment]
@sandrogauci, Then there is the case of the SRTP layer (which uses its own ciphersuites).

Current requirement is:

> 53.2.2 [ADDED] Verify that the media server is configured to use and support strong cipher suites for the Datagram Transport Layer Security (DTLS) exchange, ensuring that the selected cipher suites are considered strong and secure.

Should this be expanded to include SRTP exchanges which are technically not DTLS (but work in parallel to DTLS)?

Proposition:

> 53.2.2 [ADDED] Verify that the media server is configured to use and support strong cipher suites for the Datagram Transport Layer Security (DTLS) and SRTP exchanges, ensuring that the selected DTLS cipher suites and DTLS-SRTP protection profiles are considered strong and secure.

***

See [RFC8827](https://datatracker.ietf.org/doc/html/rfc8827):

> The DTLS-SRTP protection profile SRTP_AES128_CM_HMAC_SHA1_80 MUST be supported for SRTP. Implementations MUST favor cipher suites which support Forward Secrecy (FS) over non-FS cipher suites and SHOULD favor Authenticated Encryption with Associated Data (AEAD) over non-AEAD cipher suites. Note: the IETF is in the process of standardizing DTLS 1.3 [[TLS-DTLS13](https://tools.ietf.org/html/draft-ietf-tls-dtls13-39)].

See list of DTLS-SRTP [protection profiles](https://www.iana.org/assignments/srtp-protection/srtp-protection.xhtml).

**Comment by sandrogauci** [comment]
@randomstuff Yes it makes total sense to update the 53.2.2 requirement to include DTLS-SRTP protection files as those should also be checked.

Your updated text makes sense. Will you send out a pull request yourself or do you prefer if I do that?

**Comment by randomstuff** [comment]
OK thanks, I'll do that later on today (but if you want to make the PR in the meanwhile, no problem 😄 ).

**Comment by sandrogauci** [comment]
I'll leave it up to you then, thanks for this!

**Comment by randomstuff** [comment]
I updated the PR to simplify the wording because my previous version was not really correct and I thought it was quite redundant anyway:

> 53.2.2 [ADDED] Verify that the media server is configured to use and support strong and secure DTLS cipher suites and DTLS-SRTP protection profiles.

**Comment by tghosth** [comment]
Nice, thanks both!
